source/particles/particle_handler.cc

```diff
@@ -2193,20 +2252,25 @@ namespace Particles
           {
             // we need to find the correct child to store the particles and
             // their reference location has changed
-            const unsigned int stored_index =
-              cell_to_store_particles->active_cell_index();
+            const typename particle_container::iterator &cache =
+              cells_to_particle_cache[cell_to_store_particles
+                                        ->active_cell_index()];
 
             // Cannot use range-based loop, because number of particles in cell
             // is going to change
+            if (cache == typename particle_container::iterator())
+              break;
```

Comment:
This line confuses me, because it should never happen (and if it does something went wrong). Above in line 2216 we inserted the particle data into the new cell. In this case into child(0) of the old cell that was refined. during that operation the cache should have been updated to the correct number of new particles, so cache should never be empty. Maybe make this an Assert?

Reply:
I don't know exactly why, but when I ran one of the Lethe tests I have the case that `data_range` is empty, and hence, no entries get added. I'm not sure whether this is supposed to happen or not. Either way, I return early from the function in that case now, and changed to an assertion.